shell: add basic job shell with subprocess, environment, and PMI support

garlick opened the pull request:
(Edit: updated 4 July 2019)
(Edited to reflect the current state of this PR 3 July 2019.)

Here is a redo on #2196, a skeletal flux-shell.
```
Usage: flux-shell [OPTIONS] JOBID
  -R, --resources=FILE   Get R from FILE, not job-info service
  -h, --help             Display this message.
  -j, --jobspec=FILE     Get jobspec from FILE, not job-info service
  -r, --broker-rank=RANK Set broker rank, rather than asking broker
  -s, --standalone       Run local program without Flux instance
  -v, --verbose          Log actions to stderr
```

This closes #2186 (basic), #2187 (subproc, environ, exit codes), and #2189 (PMI).

The algorithm for mapping tasks to ranks is copied from wreck.  Specifically the `rcalc` class and its tests were brought back from the abyss and modified to accept the new, compact and versioned R_lite version 1.

The jobspec parser is patterned after libjj in the simple-sched.  It is limited to one task per slot, and resource hierarchies of node->slot->core or slot->core.  Sorry I didn't improve upon existing technology for R and jobspec.  Perhaps that can be follow-on work.

Stdio is conspicuously not implemented, so stdio from tasks is mixed with that of the shell, which job-exec logs to the flux instance.

Tasks are not bound to cores.

PMI works but has not been extensively tested.  When only one shell is involved, the KVS is not used.  When multiple shells are involved, the guest KVS namespace is used, e.g. kvs namespace looks like this for a 4 node MPICH hello:
```
job.0000.02cc.ff00.0000.J = dmVyc2lvbgBpMQB1c2VyaWQAaTU1ODgAbWVjaABzbm9uZQA=....
job.0000.02cc.ff00.0000.R = {"execution": {"R_lite": [{"children": {"core": "...
job.0000.02cc.ff00.0000.eventlog = {"timestamp":1562189451.0888035,"name":"su...
job.0000.02cc.ff00.0000.guest.exec.eventlog = {"timestamp":1562189451.1050961...
job.0000.02cc.ff00.0000.guest.pmi.P0-businesscard = description#morbo$port#44...
job.0000.02cc.ff00.0000.guest.pmi.P1-businesscard = description#morbo$port#38...
job.0000.02cc.ff00.0000.guest.pmi.P2-businesscard = description#morbo$port#44...
job.0000.02cc.ff00.0000.guest.pmi.P3-businesscard = description#morbo$port#56...
job.0000.02cc.ff00.0000.guest.pmi.hostname[0] = morbo
job.0000.02cc.ff00.0000.guest.pmi.hostname[1] = morbo
job.0000.02cc.ff00.0000.guest.pmi.hostname[2] = morbo
job.0000.02cc.ff00.0000.guest.pmi.hostname[3] = morbo
job.0000.02cc.ff00.0000.guest.pmi.sharedFilename[0] = /dev/shm/mpich_shar_tmp...
job.0000.02cc.ff00.0000.jobspec = {"attributes": {"system": {"cwd": "/home/ga...
```
(So like before, except private namespace).  I note that our jobid's (uint64_t) don't map extremely well to the (integer) PMI appnum.  I'm not sure if that's a big deal.

The `job-exec` module was modified to exec this shell instead of `/bin/true` if another one is not specified by broker attribute or jobspec.  It was also modified to set the FLUX_KVS_NAMESPACE environment variable for the shell (and the job), pointing at the guest namespace.

Upon changing the defualt job shell early on, the valgrind sharness test started failing in what looked like a module (couldn't tell which one).  I had worked around this by setting the shell in the sharness valgrind test back to /bin/true.  I just dropped the workaround to come back to it, and isn't there anymore on my development system.  We'll see about travis.  [Edit: it's back in travis.  I opened #2217 and added the workaround back to this PR]

Apologies in advance for the rush job - as usual things took longer than I thought, and to get this merged on our two-week cadence, there is inevitably some yuck in here that shouldn't be.

chu11 commented on src/shell/info.h:
```diff
@@ -0,0 +1,49 @@
+struct shell_info {
+    flux_jobid_t jobid;
+    int shell_rank;
+    int shell_size;
```
`shell_size` made me think "size of the shell", which makes no sense.  Should it really be `job_node_count` or something more like that?

garlick replied:
I struggled with that one!  I was trying to make it clear that the shells are themselves a parallel program of sorts with a rank and size.  Hmm, let me ponder some more.

garlick replied:
Er this was left as is in the current PR for now.  I'm open to changing later.

chu11 commented on src/modules/job-exec/exec.c:
```diff
@@ -41,13 +40,11 @@ static const char *jobspec_get_job_shell (json_t *jobspec)
-                  "%ju: no configured job shell, using %s",
-                  (uintmax_t) job->id, path);
-    }
+    if (!path && !(path = flux_attr_get(job->h, "job-exec.job-shell")))
```
nit: code style space before `(`

chu11 commented on src/shell/jobspec.c:
```diff
@@ -0,0 +1,160 @@
+    /* For jobspec version 1, expect exactly one array element per level.
+     */
+    if (json_unpack_ex (o, &loc_error, 0,
+                        "[{s:s s:i s?o}]",
```
what is `?` for.  Skimming API ref, `?` is only for packing not unpacking?

chu11 replied:
nevermind, I now see: `New in version 2.3: Any s representing a key may be suffixed with a ? to make the key optional. If the key is not found, nothing is extracted. See below for an example.`

garlick replied:
From https://jansson.readthedocs.io/en/2.6/apiref.html#parsing-and-validating-values

> New in version 2.3: Any s representing a key may be suffixed with a ? to make the key optional. If the key is not found, nothing is extracted. See below for an example.

garlick commented:
This is lauching tasks now, albeit without PMI or stdio support.  I hope to get some tests written tomorrow.

I still haven't run down the valgrind failure (currently working around it by adding `-o,--setattr=job-exec.job-shell=/bin/true` to the valgrind broker command line.)

garlick commented:
Restarted builder that failed here (appears unrelated)
```
ERROR: t2005-hwloc-basic.t - exited with status 137 (terminated by signal 9?)
```

garlick commented:
OK, forced a push with a couple more sharness tests that bumped up coverage to an acceptable level.  All done poking at it for today.

I'm leaving town on Fri the 5th to return on the 15th, but I'll keep one eye on github and address any egregious issues I've created with this.   I'm for merging ASAP and getting on with things though.

garlick commented:
OK, I couldn't leave it alone.  Pushing the following changes (squashed):
* drop gratuitous barrier
* add big comment block in pmi.c
* stub out stdio, stderr handling in io.c

garlick commented:
Argh, seems as though I'm randomly losing stdout from one rank in a two rank job in the standalone shell test in travis, after dropping the subprocess FLUX_SUBPROCESS_FLAGS_STDIO_FALLTHROUGH flag and adding a callback for stdout, stderr.  I must have some racy synchronization or not being using libsubprocess right.

Going to start packing up soon to leave on vacation, so I'll back that change out if I can't fix it, so I can leave this PR in a passing state.

garlick commented:
Ok, apparently the reactor can exit before all subprocess I/O has been consumed.  I added a "flush" of sorts when tasks are destroyed and now it seems all the output is showing up.  I squashed the fix and forced a push.

Possibly some review from @chu11 on the usage of the subprocess API would be useful here.  Note: at this early stage I didn't implement special handling of early task exit, etc., thus the subprocess state callback is currently not registered, but I expect it will need to be.  We should talk over what events we would like in the guest eventlog (all tasks started, first task exited, etc).

garlick commented:
Right, strike that, apparently I just got lucky on one travis run.  The tests failed after squashing the changes.  I don't have to work on this right now so I just pushed a change which disables the I/O capture for the time being.  It didn't really accomplish anything anyway - just laying groundwork for #2190.  Perhaps it should be looked at comprehensively after this goes in.

chu11 commented on src/common/libpmi/simple_server.c:
```diff
@@ -28,14 +28,13 @@
- *   the user to complete the barrier, and this callback is required.
- *   From this callback, call pmi_simple_server_barrier_complete() once the
- *   user-held count reaches 'universe_size'.
+ * barrier_enter
```
typo in the commit message "inacurate"

chu11 commented on src/shell/pmi.c:
could be non-overflow error - just `log_err ("shell_pmi_kvs_key")`?

chu11 commented on src/cmd/flux-job.c:
```diff
@@ -676,8 +677,12 @@ void attach_event_continuation (flux_future_t *f, void *arg)
+            if (json_unpack (context, "{s:i}", "status", &status) < 0)
                 log_err_exit ("error decoding finish context");
+            if (WIFSIGNALED (status))
+                ctx->exit_code = WTERMSIG (status) + 128;
```
Now that I think about it, I made an issue for it #2219

chu11 commented on t/t2601-job-shell-standalone.t:
```diff
@@ -0,0 +1,155 @@
+jq=$(which jq 2>/dev/null)
+test -n "$jq" && test_set_prereq HAVE_JQ
+
+PMINFO=${FLUX_BUILD_DIR}/src/common/libpmi/test_pminfo
```
nit: here and elsewhere, should this be "PMIINFO", with two I's in the middle?  I read this as "PM INFO" at first.

garlick commented:
Thanks for the input @chu11!  I will try to get a time slice to work on this on Wed or Thurs.

Welcome back @grondo - if you have a moment to review I'd appreciate any feedback/ideas for improvement to get this on the right footing for the next steps.

The two major outstanding problems are 1) why I lose stdio lines when capturing task I/O, and 2) what memory leak is triggered when this shell is the default for the valgrind sharness test.

grondo commented:
> if you have a moment to review I'd appreciate any feedback/ideas for improvement to get this on the right footing for the next steps.

Yes, was just starting to look at this PR along with #2208, since I didn't have any time to follow this the past 2 weeks it might take more than a moment though...

garlick commented:
That is fine - for the most part I won't be able to dig into this again until next week.

chu11 commented:
> 1) why I lose stdio lines when capturing task I/O

Just to clarify, you're losing some lines, but not all lines?  But with commit https://github.com/flux-framework/flux-core/pull/2211/commits/294cc63cb8d131ad1ac263256ed0e82722f13425 all lines output?

garlick commented:
Correct.

On Tue, Jul 9, 2019 at 10:07 AM Al Chu <notifications@github.com> wrote:

>
>    1. why I lose stdio lines when capturing task I/O
>
> Just to clarify, you're losing some lines, but not all lines? But with
> commit 294cc63
> <https://github.com/flux-framework/flux-core/commit/294cc63cb8d131ad1ac263256ed0e82722f13425>
> all lines output?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/2211?email_source=notifications&email_token=AABJPW5GCBETJMWV2DXCZJ3P6TAWJA5CNFSM4H3W7TPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZQ475A#issuecomment-509726708>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABJPW6NXPYEFGZKSSUT5ZDP6TAWJANCNFSM4H3W7TPA>
> .
>

garlick commented on src/shell/task.c:
```diff
@@ -0,0 +1,256 @@
+    int flags = 0;
+
+    if (!task->io_cb)
+        flags |= FLUX_SUBPROCESS_FLAGS_STDIO_FALLTHROUGH;
```
The intent was to (initially) fold I/O in with the shell's, for capture by the `job-exec` service, which dumps it to the flux log.  To prepare for a next step where I/O goes to your service, I added the callbacks and was intending, just for now, to have the callbacks implement the same behavior - emit to the shell's I/O, for capture by `job-exec`.

However when I did that, it seemed sometimes one task's output was lost in travis, so I left the little bit of code there but made it behave as before by adding the FALLTHROUGH flag.

I probably shouldn't have taken that list little step, but it may have uncovered some other problem with the shell terminating too soon, so it seems like a good thing to understand.

grondo replied:
I'm still trying to read and understand all the code. However, when you switched from "fallthrough" to having output handled by a callback, did you ensure to "flush" all the output from buffers of exited tasks. @chu11, correct me if I am wrong, but the "completed" callback for a subprocess only ensures EOF has been buffered on stdout/err for the process, not that it has been read or handled by a callback.

garlick replied:
I did not.  The code that failed is all there with the per-task initialization commented out.

grondo replied:
If something like the above isn't the problem, I wonder if the job-exec module could be responsible for dropping the output. Likely the bulk-exec code isn't super careful about handling all the IO. Maybe dump output to a file using the cb mechanism to see if job-exec is the culprit?

grondo replied:
Ah, just as a future note, I also noticed the callback will/might drop output if the tasks output doesn't end with a newline. I think flux-exec handles that case by checking in the output handler if a task has exited, and grabbing *all* remaining output if so.

chu11 replied:
Oops, missed these comments yesterday.

>  @chu11, correct me if I am wrong, but the "completed" callback for a subprocess only ensures EOF has been buffered on stdout/err for the process, not that it has been read or handled by a callback.

Correct.  The EOF has been "sent" to the stdout/stderr callback, but there is no guarantee that it has been read.

garlick replied:
> Ah, just as a future note, I also noticed the callback will/might drop output if the tasks output doesn't end with a newline. I think flux-exec handles that case by checking in the output handler if a task has exited, and grabbing all remaining output if so.

I did copy that code from `flux-exec` (it is in shell/task.c):

```c
/* N.B. Ref: cmd/flux-exec.c output_cb()
 */
int shell_task_io_readline (struct shell_task *task,
                            const char *name,
                            const char **line)
{
    const char *buf;
    int len;

    if (!(buf = flux_subprocess_read_line (task->proc, name, &len)))
        return -1;
    if (len == 0) {
        if (flux_subprocess_state (task->proc) == FLUX_SUBPROCESS_EXITED) {
            if (!(buf = flux_subprocess_read (task->proc, name, -1, &len)))
                return -1;
        }
    }
    *line = buf;
    return len;
}
```

I'm not sure it matters, but I am not synchronizing using the "completed" callback, but instead just letting the reactor exit naturally when subprocess unregisters all its watchers.

I have a commit waiting to try (after verifying valgrind issue is gone) that will call the I/O callback one last time after the reactor exits to consume any buffered I/O, although I'm pretty sure I tried that before and it didn't help.

grondo replied:
Ah ok, sorry I missed that. Since you have that I'm wondering again if the problem is in the job-exec module.

BTW, how were you detecting dropped lines of output?

garlick replied:
`t2601-job-shell-standalone.t` runs the job shell without `job-exec` and the missing lines of output are detected there, so `job-exec` is innocent!  Sorry, should have spoken up when you mentioned that before.

grondo replied:
> t2601-job-shell-standalone.t runs the job shell without job-exec and the missing lines of output are detected there

Hm, I uncommented `shell_task_io_enable` and don't see an error in that test
```diff
diff --git a/src/shell/shell.c b/src/shell/shell.c
index db54b19..c46e77a 100644
--- a/src/shell/shell.c
+++ b/src/shell/shell.c
@@ -209,8 +209,8 @@ int main (int argc, char *argv[])
             log_err_exit ("shell_task_create index=%d", i);
         if (shell_task_pmi_enable (task, shell_pmi_task_ready, pmi) < 0)
             log_err_exit ("shell_task_pmi_enable");
-        //if (shell_task_io_enable (task, shell_io_task_ready, io) < 0)
-        //    log_err_exit ("shell_task_io_enable");
+        if (shell_task_io_enable (task, shell_io_task_ready, io) < 0)
+            log_err_exit ("shell_task_io_enable");
         if (shell_task_start (task, r, task_completion_cb, info) < 0)
             log_err_exit ("shell_task_start index=%d", i);
```

```
~/git/flux-core.git/t$ ./t2601-job-shell-standalone.t 
ok 1 - flux-shell: generate 1-task jobspec and matching R
ok 2 - flux-shell: run 1-task echo job
ok 3 - flux-shell: -v output includes expected task count
ok 4 - flux-shell: missing JOBID fails with Usage message
ok 5 - flux-shell: missing -r fails with standalone message
ok 6 - flux-shell: missing -R fails with standalone message
ok 7 - flux-shell: missing -j fails with standalone message
ok 8 - flux-shell: nonexistent jobspec file fails
ok 9 - flux-shell: malformed jobid fails
ok 10 - flux-shell: out of range jobid fails
ok 11 - flux-shell: out of range broker rank fails
ok 12 - flux-shell: wrong range broker rank fails
ok 13 - flux-shell: unknown argument fails
ok 14 - flux-shell: generate 2-task jobspec and matching R
ok 15 - flux-shell: run 2-task printenv job
ok 16 - flux-shell: 0: completed with no error
ok 17 - flux-shell: 1: completed with no error
ok 18 - flux-shell: 0: FLUX_TASK_LOCAL_ID, TASK_RANK set
ok 19 - flux-shell: 1: FLUX_TASK_LOCAL_ID, TASK_RANK set
ok 20 - flux-shell: FLUX_JOB_SIZE, JOB_NNODES, JOB_ID set
ok 21 - flux-shell: 0: PMI_RANK set
ok 22 - flux-shell: 1: PMI_RANK set
ok 23 - flux-shell: PMI_SIZE, PMI_FD set
ok 24 - flux-shell: generate 8-task bash exit rank job
ok 25 - flux-shell: environ in jobspec is set for task
ok 26 - flux-shell: shell PMI works
ok 27 - flux-shell: shell PMI KVS works
ok 28 - flux-shell: shell can launch flux
ok 29 - flux-shell: shell exits with highest task exit value
# passed all 29 test(s)
1..29
```

chu11 replied:
I wasn't able to reproduce as well.

garlick replied:
I've only seen it in travis, sadly.

grondo replied:
Yeah, I think I see it in the current travis run. Perhaps the constrained travis container causes a race to be triggered. (though it definitely isn't obvious what the problem could be as all the code looks fine to me.)

grondo replied:
I've found I can reproduce by binding a shell in standalone mode to a single core:
```
$ taskset -c 0 ../src/shell/flux-shell -s -v -r 0 -R trash-directory.t2601-job-shell-standalone/R2 -j trash-directory.t2601-job-shell-standalone/j2 123456 > j2.out 
flux-shell: 0: task_count=2 slot_count=2 cores_per_slot=1 slots_per_node=2
flux-shell: 0: tasks [0-1] on cores 0-1
0: C: pmi EOF
flux-shell: task 0 complete status=0
1: C: pmi EOF
flux-shell: task 1 complete status=0
flux-shell: exit 0
$ grep -c ^0: j2.out
109
$ grep -c ^1: j2.out
36
```

(similarly, running `t2601-job-shell-standalone.t` under `taskset -c 0` reproduces the travis failures) Maybe that will help find root cause?

garlick replied:
Oooh, nice!   I can't seem to reproduce it on my virtualbox vm with that (and anyway it's only got one core).  Maybe I need to fire up some different hardware here to get it to happen.

grondo replied:
I added some printf debug to `shell_task_io_readline` and the results are a bit strange. 

```diff
diff --git a/src/shell/task.c b/src/shell/task.c
index 16b48e3..06e32de 100644
--- a/src/shell/task.c
+++ b/src/shell/task.c
@@ -148,10 +151,12 @@ int shell_task_io_readline (struct shell_task *task,
 
     if (!(buf = flux_subprocess_read_line (task->proc, name, &len)))
         return -1;
+    fprintf (stderr, "%d: %s: read %d bytes\n", task->rank, name, len);
     if (len == 0) {
         if (flux_subprocess_state (task->proc) == FLUX_SUBPROCESS_EXITED) {
             if (!(buf = flux_subprocess_read (task->proc, name, -1, &len)))
                 return -1;
+            fprintf (stderr, "%d: %s: after exit: %d bytes\n", task->rank, name, len);
         }
     }
     *line = buf;
```

This might not be directly related to the case of the missing lines though

```
[snip a bunch of 0: STDOUT lines]

0: STDOUT: read 16 bytes
0: STDOUT: read 18 bytes
0: STDOUT: read 19 bytes
0: STDOUT: read 11 bytes
0: STDOUT: read 11 bytes
0: STDOUT: read 10 bytes
0: STDOUT: read 0 bytes
0: STDOUT: after exit: 0 bytes
0: STDOUT: read 0 bytes
0: STDOUT: after exit: 0 bytes
0: STDERR: read 0 bytes
0: STDERR: after exit: 0 bytes
flux-shell: task 0 complete status=0
1: STDOUT: read 19 bytes
1: STDOUT: read 9 bytes
1: STDOUT: read 85 bytes
1: STDOUT: read 12 bytes
1: STDOUT: read 1421 bytes
1: STDOUT: read 15 bytes
1: STDOUT: read 122 bytes
1: STDOUT: read 15 bytes
1: STDOUT: read 25 bytes
1: STDOUT: read 38 bytes
1: STDOUT: read 45 bytes
1: STDOUT: read 15 bytes
1: STDOUT: read 330 bytes
1: STDOUT: read 218 bytes
1: STDOUT: read 11 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: read 0 bytes

// above repeats about 3000x
1: STDOUT: read 0 bytes
1: STDOUT: after exit: 1 bytes
1: STDOUT: read 0 bytes
1: STDOUT: after exit: 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: after exit: 0 bytes
1: C: pmi EOF
1: STDERR: read 0 bytes
1: STDERR: after exit: 0 bytes
flux-shell: task 1 complete status=0
0: STDOUT: read 0 bytes
0: STDOUT: after exit: 0 bytes
0: STDERR: read 0 bytes
0: STDERR: after exit: 0 bytes
0: STDOUT: read 0 bytes
0: STDOUT: after exit: 0 bytes
0: STDERR: read 0 bytes
0: STDERR: after exit: 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: after exit: 0 bytes
1: STDERR: read 0 bytes
1: STDERR: after exit: 0 bytes
1: STDOUT: read 0 bytes
1: STDOUT: after exit: 0 bytes
1: STDERR: read 0 bytes
1: STDERR: after exit: 0 bytes
flux-shell: exit 0
```

I'm not sure this tells us that much. However a couple points:

 - It will save a lot of useless looping if line buffered streams aren't "ready" for reading unless at least one line is available to read.

 - I'm not sure how EOF handling works with local subprocess buffered I/O. I don't see where an EOF is marked as pending or consumed on streams (obviously this must work somehow, or have we just not dealt with local EOFs yet?)

Other than that, it looks like task 1 buffer just stops reading data after some number of lines. Not sure why yet. (sorry for the inconclusive data post, but figured maybe @chu11 might have some ideas where to look next)

grondo replied:
Another hint, added a printf line in `subprocess_read` to display the size of the buffer we're reading from and on a failing run I see, for task 0:
```
subprocess_read: STDOUT: lines=109 bytes=8767
0: STDOUT: read 36 bytes
subprocess_read: STDOUT: lines=108 bytes=8731
0: STDOUT: read 278 bytes
...
```

whereas task 1 shows:

```
flux-shell: task 0 complete status=0
subprocess_read: STDOUT: lines=35 bytes=4096
1: STDOUT: read 36 bytes
subprocess_read: STDOUT: lines=34 bytes=4060
1: STDOUT: read 278 bytes
...
flux-shell: task 0 complete status=0
subprocess_read: STDOUT: lines=35 bytes=4096
1: STDOUT: read 36 bytes
subprocess_read: STDOUT: lines=34 bytes=4060
1: STDOUT: read 278 bytes
```

Suspciously task 1 buffer is only filled to 4K bytes and never gets any more data...

chu11 replied:
> It will save a lot of useless looping if line buffered streams aren't "ready" for reading unless at least one line is available to read.

This seems like a useful option to have for `libsubprocess`.  Right now the watcher is not line buffered.  Should be easy to add with `FLUX_WATCHER_LINE_BUFFER`.

> 1: STDOUT: read 0 bytes
> 1: STDOUT: read 0 bytes
> 1: STDOUT: read 0 bytes

my initial thought was if there was some bug where the watcher was not shut down properly and constantly calling the `on_stdout` callback.  But it appears that is shut off once EOF is sent to the caller.

Are you reproducing this via the `taskset` reproducer you have above?

grondo replied:
> Are you reproducing this via the taskset reproducer you have above?

Yeah, the simplest reproducer is to first run `./t2601-job-shell-standalone -d -v` to get the appropriate input files in `trash-directory.t2601-job-shell-standalone`, then run:

```
 taskset -c 0 ../src/shell/flux-shell -s -v -r 0 -R trash-directory.t2601-job-shell-standalone/R2 -j trash-directory.t2601-job-shell-standalone/j2 123456 > j2.out 2>log
```

On fluke, this seems to reproduce the issue 100%.

I'm actually wondering if we should try to make a more isolated reproducer that just runs two local subprocesses to take the job-shell out of the equation. Something funky seems to be going on with buffered io here. In my testcase I added some printf debug down in the ev_buffer_read code and I'm not seeing where the subprocess buffers are even reading EOF from their stdout streams...

```console
$ taskset -c 0 ../src/shell/flux-shell -s -v -r 0 -R trash-directory.t2601-job-shell-standalone/R2 -j trash-directory.t2601-job-shell-standalone/j2 123456 > j2.out 2>log
$ grep buffer_read_cb log
buffer_read_cb: fd=13, space=1048576, read=0 bytes
buffer_read_cb: fd=13. eof_read=true
buffer_read_cb: fd=9, space=1048576, read=8767 bytes
buffer_read_cb: fd=11, space=1048576, read=0 bytes
buffer_read_cb: fd=11. eof_read=true
buffer_read_cb: fd=12, space=1048576, read=4096 bytes
buffer_read_cb: fd=17, space=1048576, read=0 bytes
buffer_read_cb: fd=17. eof_read=true
buffer_read_cb: fd=15, space=1048576, read=0 bytes
buffer_read_cb: fd=15. eof_read=true
```

Note no `eof_read=true` for fds 9 and 12, the stdout fds of task 0 and 1 respectively.

grondo replied:
Note I see the same issue with no eof on the non-failing reproducer (without `taskset -c 0`), however the difference is that the initial `read` into the buffer for task 1 is the full `8767` bytes.

Makes me wonder if we make a test to output much larger files if data will be dropped even in the non `taskset -c0` case.

chu11 replied:
Ok, got it reproduced (took me awhile to realize, oh I should probably be in a flux instance).  Reproducing on catalyst and fluke consistently.  Can dig into this deeper tomorrow.

chu11 replied:
So I added some debugging to the buffer reactor (libflux/ev_buffer_read.c)

> buffer_read_cb:96 write=4096
> subprocess_read: STDOUT: lines=89 bytes=4096
> 1: STDOUT: read 26 bytes

AFAICT, the task is only writing 4096 bytes, so apparently only 4096 bytes of data are coming
from task 1.  So that's an interesting nugget.

grondo replied:
> AFAICT, the task is only writing 4096 bytes, so apparently only 4096 bytes of data are coming from task 1.

I think it unlikely that task 1 is only writing 4096 bytes. I think I did try running something other than `printenv` which had guaranteed same output from each task and still some of task 1 stdout was dropped.

chu11 replied:
Sorry, what I meant was "AFAICT, the task is only writing 4096 bytes instead of the (> 4096) bytes it should be."  

I forced libsubprocess to be line buffered and the reproducer started working.  Saw:

> buffer_read_cb:96 write=4096
> buffer_read_cb:96 write=247

in task 1, which is equal to task 0's

> buffer_read_cb:96 write=4343

I suppose task 1 isn't `fflush`-ing or something like task 0, but somehow on the second write
isn't completing.

grondo replied:
Ah, good hint! However, shouldn't a process exiting force stdio flush?

chu11 replied:
> However, shouldn't a process exiting force stdio flush?

I would think so.

garlick replied:
Looks like an `atexit()` handler flushes buffered stdio when `exit()` is called:
https://www.gnu.org/software/libc/manual/html_node/Flushing-Buffers.html

chu11 replied:
so I think I have a lead on the problem.  There seems to be a circumstance where the `flux_buffer_read_watcher()` does NOT read EOF on a stream, BUT `libsubprocess` calculates that EOF has occurred and sends EOF to a `libsubprocess` callback. 

I'm zero-ing on this part of the code in libsubprocess in `local.c`.  I'm thinking that the `output_cb` (which calls the user's `on_stdout`) reads all the data in the buffer.  The child process has exited, BUT all of the stdout hasn't been flushed / read into the buffer yet.  So this is not a safe calculation of EOF.  Will have to look into this more tomorrow and figure out a reproducer, tests, and how to fix.  And if the problem exists for non-local subprocesses too.

I hope this is it, this is a bit of a doozy :-)

```
        output_cb (c->p, c->name);

        if (c->p->state == FLUX_SUBPROCESS_EXITED && !c->eof_sent_to_caller) {

            if (!(fb = flux_buffer_read_watcher_get_buffer (w))) {
                flux_log_error (c->p->h, "flux_buffer_read_watcher_get_buffer");
                return;
            }

            if (!flux_buffer_bytes (fb)) {

                output_cb (c->p, c->name);

                c->eof_sent_to_caller = true;
                eof_set = true;
                c->p->channels_eof_sent++;
            }
        }

```

grondo replied:
Nice sleuthing! That looks promising to me! We could probably reproduce by running a local subprocess with a lot of output (so that child exit is detected before all output is complete), and maybe shrinking the childs stdio buffer size (i.e. via setvbuf()) might make this situation more likely to occur. 

BTW, should we move this to its own issue (and copy the last few comments here)?

chu11 replied:
Tomorrow I'll try to 100% confirm it.  Many many random printf's all over the place, want to try and confirm it is specifically this and that `ev_io()` and the `flux_buffer_read_watcher()` work as I believe it should.

chu11 replied:
is now issue #2228

grondo commented:
> what memory leak is triggered when this shell is the default for the valgrind sharness test.

For you, was this issue only reproducible under the centos7 travis docker image? I can't reproduce an issue anywhere else. Maybe a false positive?

garlick commented:
No it popped up in Ubuntu as well.

On Tue, Jul 9, 2019, 2:26 PM Mark Grondona <notifications@github.com> wrote:

> what memory leak is triggered when this shell is the default for the
> valgrind sharness test.
>
> For you, was this issue only reproducible under the centos7 travis docker
> image? I can't reproduce an issue anywhere else. Maybe a false positive?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/2211?email_source=notifications&email_token=AABJPW5E3B4TLADSU4HM55DP6T7ALA5CNFSM4H3W7TPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZRSX2A#issuecomment-509815784>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABJPWY3CMS3TQLVQDGVM7LP6T7ALANCNFSM4H3W7TPA>
> .
>

grondo commented:
> No it popped up in Ubuntu as well. 

I've been unable to reproduce on Ubuntu.

Also, the errors I was getting in the Centos 7 docker image were because the `valgrind-devel` package was missing, and thus our valgrind wrap of `dlclose()` wasn't active. (I'll update the docker testenv image to add this I think)

garlick commented:
Now I'm wondering if I am confusing this with the io failures...

On Tue, Jul 9, 2019, 4:40 PM Mark Grondona <notifications@github.com> wrote:

> No it popped up in Ubuntu as well.
>
> I've been unable to reproduce on Ubuntu.
>
> Also, the errors I was getting in the Centos 7 docker image were because
> the valgrind-devel package was missing, and thus our valgrind wrap of
> dlclose() wasn't active. (I'll update the docker testenv image to add
> this I think)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/2211?email_source=notifications&email_token=AABJPW3LRY5VKNEDL4COZZLP6UOWJA5CNFSM4H3W7TPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZR3AUQ#issuecomment-509849682>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABJPW2KCMVLEOQYBBX2DQ3P6UOWJANCNFSM4H3W7TPA>
> .
>

grondo commented:
FWIW, I also uncommented the `shell_task_io_enable()` lines and still don't get any leaks from the valgrind test, even when I run jobs that produce a line of output per task (`hostname`).

garlick commented:
Just pushed a few commits to address review comments thus far, except for the commit message spelling error (will catch that when we squash/rebase).  I backed out the change to the valgrind test since @grondo updated the centos docker images.  I haven't changed anything yet to flush I/O on task completion (still need to study that a bit).

garlick commented:
Restarted centos 7 python 3 builder after docker image fix.

grondo commented:
> Restarted centos 7 python 3 builder after docker image fix.

Unfortunately centos7 python 3 builds will be broken until #2221 is merged, since we had to update python test version from 3.4 to 3.6. Sorry.

garlick commented:
Oops.  Okay, merged that.  Any objection if I rebase this on current master?

grondo commented:
No rebase away

garlick commented:
Restarted ubuntu builder that failed valgrind test due to #2195

grondo commented on t/t2601-job-shell-standalone.t:
```diff
@@ -0,0 +1,155 @@
+	grep -q FLUX_JOB_SIZE=2 printenv.out &&
+	grep -q FLUX_JOB_NNODES=1 printenv.out &&
+	grep -q FLUX_JOB_ID=42 printenv.out &&
+	! grep -q FLUX_URI printenv.out
```
The test for !`FLUX_URI` should be split into a different test here, IMO -- since the test could fail even though `FLUX_JOB_SIZE` `JOB_NNODES` and `JOB_ID` are all set (which could confuse the viewer)

Also, I noticed that this test fails when the standalone test is run under an existing instance because of this code in `shell_task_create`:
```C
    flux_cmd_unsetenv (task->cmd, "FLUX_URI");
    if (getenv ("FLUX_URI")) {
        if (flux_cmd_setenvf (task->cmd, 1, "FLUX_URI", "%s",
                              getenv ("FLUX_URI")) < 0)
            goto error;
    }
```

If a `FLUX_URI` is already set in the user's environment (which is the case when running under an instance or probably in future on a system with a Flux system instance), then even in standalone mode the shell will propagate `FLUX_URI`.

To fix, you could force unset FLUX_URI during the test.

grondo replied:
Additionally, I'd suggest removal of all `-q` arguments to `grep` in these tests. sharness already suppresses stdout when tests are run normally, and the `-q` makes it harder to debug the tests when run with `-d -v`...

garlick commented:
I went ahead and cherry picked the fix from #2230 to see how we do in travis.

garlick commented:
Hmm, after cherry picking that fix, `t2602-job-shell.t` is failing (so far just in travis)
```
expecting success: 
        id=$(flux jobspec srun -N4 bash -c \
            "flux kvs put test1.\$FLUX_TASK_RANK=\$FLUX_TASK_LOCAL_ID" \
            | flux job submit) &&
        flux job wait-event $id clean &&
        kvsdir=$(flux job id --to=kvs $id).guest &&
        test $(flux kvs get ${kvsdir}.test1.0) = 0 &&
        test $(flux kvs get ${kvsdir}.test1.1) = 0 &&
        test $(flux kvs get ${kvsdir}.test1.2) = 0 &&
        test $(flux kvs get ${kvsdir}.test1.3) = 0

1562979007.422975 clean
flux-kvs: job.0000.0004.7200.0000.guest.test1.0: No such file or directory
./t2602-job-shell.t: 7: test: =: unexpected operator
not ok 2 - job-shell: execute across all ranks

expecting success: 
        id=$(flux jobspec srun -N4 -n8 bash -c \
            "flux kvs put test2.\$FLUX_TASK_RANK=\$FLUX_TASK_LOCAL_ID" \
            | flux job submit) &&
        flux job wait-event $id clean &&
        kvsdir=$(flux job id --to=kvs $id).guest &&
        test $(flux kvs get ${kvsdir}.test2.0) = 0 &&
        test $(flux kvs get ${kvsdir}.test2.1) = 1 &&
        test $(flux kvs get ${kvsdir}.test2.2) = 0 &&
        test $(flux kvs get ${kvsdir}.test2.3) = 1 &&
        test $(flux kvs get ${kvsdir}.test2.4) = 0 &&
        test $(flux kvs get ${kvsdir}.test2.5) = 1 &&
        test $(flux kvs get ${kvsdir}.test2.6) = 0 &&
        test $(flux kvs get ${kvsdir}.test2.7) = 1

1562979008.019065 clean
flux-kvs: job.0000.0006.7f00.0000.guest.test2.0: No such file or directory
./t2602-job-shell.t: 7: test: =: unexpected operator
not ok 3 - job-shell: execute 2 tasks per rank
```

chu11 commented:
:'( Just to make sure, these are new?  You didn't see these before?  I'm guessing some EOF pattern was disrupted that I don't see, ugh.

garlick commented:
Correct.

garlick commented:
I made a couple changes to the `t2602-job-shell.t` test so that we can get a little more info out of the failure.

It's strange - the test jobs emit nothing on stdio. They set some values in the KVS based on environment variables, then verify that the values are correct after the jobs finish.  The jobs finish normally (eventlog says status=0) but the KVS is missing the values, as though the subprocesses never actually ran.

Aside: after messing up a hereis doc in my test, I found two other unrelated tests that had runaway hereis docs and fixed them.

grondo commented:
Strange. I pushed this branch to my own fork and I don't see the same failure.

(I do see failure in chain-lint test due to missing `&&` after the `sort >test*.exp <<-EOT`, but non-chain-lint builders succeed)

I wonder if we should try clearing all caches and rebuilding?

garlick commented:
Interesting!  I'm not seeing the job-shell test failure in my personal travis build either!  (Wasn't paying attention to it).

OK, I deleted the cache for this PR and pushed fixes for the chain-lint errors I introduced.  Sigh.  Let's see what happens.

garlick commented:
Building clean now in my personal travis but failed again here.  I'll try again with all caches deleted, not just the PR cache...

garlick commented:
That seems to have failed.  I deleted the caches in my personal travis and rebuilt there, but it still succeeded.

grondo commented:
There are some differences between a travis PR build vs a branch build, though I can't imaging how they'd have an effect here. This is a bit troubling...

Mind if I open a dummy PR with your branch to see if this problem also reproduces there, and use it for debugging? (We might have to pull out the old ssh login trick if I can remember how to enable it)

garlick commented:
Please do!

On Sat, Jul 13, 2019 at 2:44 PM Mark Grondona <notifications@github.com>
wrote:

> There are some differences between a travis PR build vs a branch build,
> though I can't imaging how they'd have an effect here. This is a bit
> troubling...
>
> Mind if I open a dummy PR with your branch to see if this problem also
> reproduces there, and use it for debugging? (We might have to pull out the
> old ssh login trick if I can remember how to enable it)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/2211?email_source=notifications&email_token=AABJPW2FVXIGL4GE4466BDLP7JECXA5CNFSM4H3W7TPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZ3Z5QA#issuecomment-511155904>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABJPW6T6LIEFUXGNUIZG6DP7JECXANCNFSM4H3W7TPA>
> .
>

grondo commented:
The travis failure discussed above seems to be related to `FLUX_KVS_NAMESPACE`. Since `flux jobspec` now propagates the current environment via jobspec, the value for `FLUX_KVS_NAMESPACE` was no longer inherited by the job tasks from the job shell, and the kvs puts went into the main kvs namespace.

This patch seems to fix it:
```diff
diff --git a/src/shell/task.c b/src/shell/task.c
index 16b48e37e..c492d844b 100644
--- a/src/shell/task.c
+++ b/src/shell/task.c
@@ -122,6 +122,11 @@ struct shell_task *shell_task_create (struct shell_info *info,
                               getenv ("FLUX_URI")) < 0)
             goto error;
     }
+    if (getenv ("FLUX_KVS_NAMESPACE")) {
+        if (flux_cmd_setenvf (task->cmd, 1, "FLUX_KVS_NAMESPACE", "%s",
+                              getenv ("FLUX_KVS_NAMESPACE")) < 0)
+            goto error;
+    }
     return task;
 error:
     shell_task_destroy (task);
```

There's a new mystery though -- how did this work in all cases outside of travis? Will look into that next.

grondo commented:
> There's a new mystery though -- how did this work in all cases outside of travis? Will look into that next.

This is obvious in retrospect. Travis PR builds work by merging the PR branch into master. So even a PR branch which is not up to date with master gets the most recent commits from master.

Since this PR is out of date with the most recent master, it doesn't include changes to add environment to jobspec by default, which is why tests succeed on personal branches and local builds.

garlick commented:
Really?  So there is an implicit merge there?  Wow, that is unexpected.
Good job figuring that out!

On Sat, Jul 13, 2019, 9:17 PM Mark Grondona <notifications@github.com>
wrote:

> There's a new mystery though -- how did this work in all cases outside of
> travis? Will look into that next.
>
> This is obvious in retrospect. Travis PR builds work by merging the PR
> branch into master. So even a PR branch which is not up to date with master
> gets the most recent commits from master.
>
> Since this PR is out of date with the most recent master, it doesn't
> include changes to add environment to jobspec by default, which is why
> tests succeed on personal branches and local builds.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/flux-framework/flux-core/pull/2211?email_source=notifications&email_token=AABJPW3WK2AFMBIFIA6TL73P7KSGHA5CNFSM4H3W7TPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZ352LA#issuecomment-511171884>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABJPW7JHVFU2LET7DUMWDDP7KSGHANCNFSM4H3W7TPA>
> .
>

grondo commented:
> So there is an implicit merge there?  Wow, that is unexpected.

It makes sense if you think about it. When testing a PR, you want to test the code as if the PR was merged to master (or a reasonable equivalent anyway). Otherwise a PR could pass CI, but fail on the master branch after it is merged.

See [here](https://docs.travis-ci.com/user/pull-requests/#how-pull-requests-are-built) for reference:

> Rather than build the commits that have been pushed to the branch the pull request is from, we build the merge between the source branch and the upstream branch.

garlick commented:
That does make sense.  I guess I'm more shocked that I didn't know that.

I'll go ahead and push that change here.  Should I rebase/squash or would that be disruptive at this point?

grondo commented:
> Should I rebase/squash or would that be disruptive at this point?

Feel free to rebase and squash!

garlick commented:
(Force pushed one extra time to address commit message misspelling that @chu11 commented on above.)

garlick commented:
Add a couple commits to back out code that shouldn't be necessary with the subprocess I/O EOF calculation fix, and make one test of env handling a little lest complex now that `flux jobspec srun --export VAR=VAL` is available.

Should be inching closer to mergeability IMHO.

garlick commented:
Looks like that's passing so I'll squash and rebase once #2230 merged.

garlick commented:
Squashed and rebased, and annotated commits to close relevant issues upon merge.

garlick commented:
Thanks!  That takes a load off.
